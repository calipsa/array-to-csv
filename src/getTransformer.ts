import {
  ColumnObjectOptional,
  ColumnObject,
  Column,
  Obj,
  ColumnKeys,
  ColumnLabels,
} from './types'

const completeObject = (obj: ColumnObjectOptional): ColumnObject => ({
  key: obj.key,
  label: obj.label || obj.key,
  transform: obj.transform || String,
})

const simpleToObject = (key: string): ColumnObject => ({
  key,
  label: key,
  transform: String,
})

const normalizeColumns = (c: Column) =>
  typeof c === 'object'
    ? completeObject(c)
    : simpleToObject(c)

export default <C extends Column>(columns: readonly C[]) => {
  const cols = columns.map(normalizeColumns)
  return (item: Readonly<Obj<ColumnKeys<C>>>) => {
    // @ts-ignore
    const o: Record<ColumnLabels<C>, string> = {}
    for (const { key, label, transform } of cols) {
      // @ts-ignore
      const val = item[key]
      // @ts-ignore
      o[label] = transform ? transform(val) : val
    }
    return o
  }
}
