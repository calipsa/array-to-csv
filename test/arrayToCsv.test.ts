import arrayToCsv from '../src'
import getTransformer from '../src/getTransformer'

import list from './data/list'

function formatDate(dateStr: any) {
  return dateStr
    ? new Date(dateStr).toUTCString()
    : ''
}

const columns = [
  {
    key: 'name',
    label: 'Name',
  },
  'parent',
  {
    key: 'createdAt',
    label: 'Created at (UTC)',
    transform: formatDate,
  },
  'count',
] as const

const toCsv = arrayToCsv(columns)

describe('arrayToCsv', () => {
  it('should transform correctly', () => {
    const item = list[0]
    const transform = getTransformer(columns)
    const result = transform(item)
    expect(result).toMatchSnapshot()
  })

  it('should convert to csv correctly', () => {
    const result = toCsv(list)
    expect(result).toMatchSnapshot()
  })
})
