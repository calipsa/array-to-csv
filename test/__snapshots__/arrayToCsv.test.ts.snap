// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arrayToCsv should convert to csv correctly 1`] = `
"Name,parent,Created at (UTC),count
cam3,site_random,\\"17 Jun 2019, 13:20\\",11
test 2,1foooooobgar1,,0
view,1foooooobgar1,,41
DCV N1,goooo parent,,50
DCV N2,E000000000000090,\\"17 Jun 2019, 13:21\\",81
DCV N1,E000000000000090,\\"14 Jun 2019, 14:17\\",6
DCV N2,E000000000000090,\\"14 Jun 2019, 14:17\\",6
foo,üñiçøď€,\\"24 Jun 2019, 15:00\\",4
sdfsdf,siteo,\\"06 Jun 2019, 17:00\\",5"
`;

exports[`arrayToCsv should transform correctly 1`] = `
Object {
  "Created at (UTC)": "17 Jun 2019, 13:20",
  "Name": "cam3",
  "count": "11",
  "parent": "site_random",
}
`;
